ruby/red-arrow/ext/arrow/record_batch.cc

```diff
@@ -0,0 +1,610 @@
+        auto field = type->child(i);
+        auto& name = field->name();
+        field_names_.push_back(rb_str_new_cstr(name.c_str()));
+      }
+      return Qnil;
+    });
+    return Status::OK();
+  }
+
+  VisitValueFunc visit_value_func_;
+  std::vector<VALUE> field_names_;
+  int field_index_;
+};
+
+class RawRecordsBuilder : public ArrayConverter {
+ public:
+  RawRecordsBuilder(VALUE records, int num_columns, bool convert_decimal=false)
+      : ArrayConverter(convert_decimal), records_(records), num_columns_(num_columns) {}
+
+  Status Add(const arrow::RecordBatch& record_batch) {
+    const int num_columns = record_batch.num_columns();
+    for (int i = 0; i < num_columns; ++i) {
+      auto array = record_batch.column(i);
+      column_index_ = i;
+      RETURN_NOT_OK(array->Accept(this));
+    }
+    return Status::OK();
+  }
+
+ protected:
+  using ArrayConverter::Visit;
+
+  Status Visit(const arrow::StructArray& array) override {
+    StructArrayConverter converter(
+        convert_decimal_,
+        [&](const int64_t row_index, VALUE key, VALUE val) {
+          VALUE hash = GetValue(row_index);
+          (void)rb::protect([&]{
+            if (NIL_P(hash)) {
+              hash = rb_hash_new();
+              VisitValue(row_index, hash);
+            }
+            return rb_hash_aset(hash, key, val);
+          });
+          return Status::OK();
+        });
+    return converter.Convert(array);
+  }
+
+ private:
+  inline VALUE GetValue(const int64_t row_index) {
+    return rb::protect([&] {
+      VALUE record = rb_ary_entry(records_, row_index);
+      if (NIL_P(record)) return Qnil;
+      return rb_ary_entry(record, column_index_);
+    });
+  }
+
+  Status VisitValue(const int64_t row_index, VALUE val) override {
+    (void)rb::protect([&] {
+      VALUE record = rb_ary_entry(records_, row_index);
+      if (NIL_P(record)) {
+        record = rb_ary_new_capa(num_columns_);
+        rb_ary_store(records_, row_index, record);
+      }
+      rb_ary_store(record, column_index_, val);
+      return Qnil;
+    });
+    return Status::OK();
+  }
+
+  // Destination for converted records.
+  VALUE records_;
+
+  // The current column index.
+  int column_index_;
+
+  // The number of columns.
+  const int num_columns_;
+};
+
+}  // namespace
+
+VALUE
+record_batch_raw_records(int argc, VALUE* argv, VALUE obj)
+{
+  VALUE kwargs;
+  bool convert_decimal = true;
+
+  rb_scan_args(argc, argv, ":", &kwargs);
+  if (!NIL_P(kwargs)) {
+    enum { k_convert_decimal, num_kwargs };
+    static ID kwarg_keys[num_kwargs];
+    VALUE kwarg_vals[num_kwargs];
+    kwarg_keys[k_convert_decimal] = rb_intern("convert_decimal");
+
+    rb_get_kwargs(kwargs, kwarg_keys, 0, num_kwargs, kwarg_vals);
+
+    if (kwarg_vals[k_convert_decimal] != Qundef && !NIL_P(kwarg_vals[k_convert_decimal])) {
+      convert_decimal = RTEST(kwarg_vals[k_convert_decimal]);
+    }
+  }
+
+  if (convert_decimal) {
+    rb_require("bigdecimal");
+  }
```

Comment:
We can assume that `bigdecimal` is always required.

Reply:
Fixed.